### PR TITLE
Conditional compilation for sanitizers

### PR DESCRIPTION
**Pull request by tmiasko**
Configure sanitize option when compiling with a sanitizer to make
it possible to execute different code depending on whether given
sanitizer is enabled or not.

### CODE REVIEW

**Comment by parched** [comment]
Would this work if you have more than one sanitizer enabled? For example, address and leak sanitizer can be used together (although I'm not sure if rustc currently supports that).

**Comment by tmiasko** [comment]
It is possible to configure multiple key-value pairs with the same key, so if
rustc were to support multiple sanitizers at the same time it should be non-issue.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@tmiasko Can you please resolve the merge conflicts?
 @GuillaumeGomez  is this ready for review?

Thank you

**Comment by tmiasko** [comment]
Rebased.

File: src/doc/unstable-book/src/language-features/cfg-sanitize.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# `cfg_sanitize`
+
+The tracking issue for this feature is: None
+
+------------------------
+
+The `cfg_sanitize` feature makes it possible to execute different code
+depending on whether a particular is enabled or not.
```
Comment (oli-obk) [inline src/doc/unstable-book/src/language-features/cfg-sanitize.md]: a particular ... ?

File: src/librustc/session/config.rs [hunk 2]
```diff
@@ -1532,6 +1532,15 @@ pub fn default_configuration(sess: &Session) -> ast::CrateConfig {
             }
         }
     }
+    if let Some(s) = &sess.opts.debugging_opts.sanitizer {
+        let s = match s {
```
Comment (oli-obk) [inline src/librustc/session/config.rs]: This could be a method on `Sanitizer`

**Review by oli-obk — CHANGES_REQUESTED** [review]
The implementation looks fine to me. A tracking issue with some links to related work around sanitizers would be good though

File: src/librustc/session/config.rs (hunk 2 above)
Comment (tmiasko) [reply src/librustc/session/config.rs]: Replaced with display trait implementation.

**Comment by tmiasko** [comment]
I used https://github.com/rust-lang/rust/issues/39699 as a tracking issue.

**Comment by tmiasko** [comment]
Updated the tests to run them only on linux x86-64, since no other target supports all sanitizers.

**Comment by tmiasko** [comment]
Rebased.

**Comment by RalfJung** [comment]
> It is possible to configure multiple key-value pairs with the same key, so if
rustc were to support multiple sanitizers at the same time it should be non-issue.

Is there precedent for that, i.e., is there any existing `cfg` flag that uses this? It seems quite counter-intuitive that `cfg(sanitizer = "a")` and `cfg(sanitizer = "b")` can both be true. That looks like it implies `"a" = "b"`.

**Comment by tmiasko** [comment]
A few examples: `feature = "name"` (configured by cargo), `target_feature = "name"`, `target_has_atomic = "size"`.

**Comment by RalfJung** [comment]
Hm, you are right. I guess that ship has sailed. Never mind then!

**Comment by RalfJung** [comment]
@bors retry
rolled up

**Pull request merged**